ideas/apq.h

```diff
@@ -0,0 +1,362 @@
+	};
+
+	detail::async_result_init<CompletionToken, void (error_code)> init(std::move(token));
+	
+	get_connection(provider, [fwd, token = init.handler](error_code ec, connection conn) {
+		if (ec) {
+			return fwd(std::move(token), ec, {});
+		}
+		basic_protocol::execute(conn, queries::begin_transaction, 
+				[fwd, conn, token = std::move(token)] (error_code ec) {
+			if (ec) {
+				return fwd(std::move(token), ec, {});
+			}
+			fwd(std::move(token), {}, {conn});
+		});
+	});
+
+	return init.result.get();
+}; // Begin transaction
+
+
+make_query( "SELECT name FROM my_table WHERE id="_sql << id << " AND sid=" << sid );
+
+make_query( "SELECT name FROM my_table WHERE id=:id  AND sid=:sid"_sql % (("id",id), ("sid",sid)) );
+
+
+BOOST_FUSION_DEFINE_STRUCT((), (my_query_params), 
+	(std::string, id)
+	(int, sid)
+);
+
+/**
+* Query from text and params
+*/
+using my_query = query<"SELECT name FROM my_table WHERE id=:id  AND sid=:sid"_sql, my_query_params>;
+
+/**
+* Query for reading from query.conf
+*/
+using my_query = query<"my_query"_query, my_query_params>;
+
+enum class debug_mode {off, on};
+
+/******************************************************
+* Minimal query interface definition.
+*******************************************************/
+
+/**
+* Function returns text part of the query
+*/
+template <typename Query>
+inline auto query_text(const Query& q) {
+    return q.text();
+}
+
+/**
+* Returns an adapted sequence (or tuple) with values
+*/
+template <typename Query>
+inline auto query_values(const Query& q) {
+    return q.values();
+}
+
+/**
+* returns query execution time limitation
+*/
+template <typename Query>
+inline std::chrono::steady_clock::duration query_timeout(const Query& q) {
+    return q.timeout();
+}
+
+/**
+* Returns query execute strategy - the object to determine execution and
+* retries. May be added later. But this must be one of the key features.
+*/
+template <typename Query>
+inline auto query_execute_strategy(const Query& q) {
+    return q.strategy();
+}
+
+
+/***********************************************
+* Type mapping traits
+***********************************************/
+template <typename T>
+struct type_traits;
+
+using oid_t = int64_t; // ????
+
+template <typename T>
+constexpr auto get_type_traits(const T&) noexcept {
+    return type_traits<std::decay_t<T>>{};
+}
+
+template <typename T>
+constexpr auto type_name(const T& v) noexcept {
+    return get_type_traits(v)::name();
+}
+
+template <typename T>
+constexpr std::size_t type_size(const T& v) noexcept {
+    return get_type_traits(v)::size;
+}
+
+// Overload trait for std::string for example
+template <>
+inline std::size_t type_size(const std::string& v) noexcept {
+    return v.size();
+}
+
+/**
+* Oid acquisition for types must looks like (simplified):
+* auto oid = type_oid(v, [oid_map](const auto& v) {
+*         return oid_map.at(type_name(v));
+*     });
```

Comment:
This case can be built with a compile-time check. It requires preparing something like boost::fusion::map with a key by type. So values of such "map" can be filled during connecting to database, or on error with invalidated oid.